apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/ElasticApmApiInstrumentation.java

```diff
@@ -86,14 +86,12 @@ private static void doStartTransaction(@Advice.Origin Class<?> clazz,
                                                @Advice.Argument(2) MethodHandle getAllHeaders,
                                                @Advice.Argument(3) @Nullable Object headersExtractor) {
             if (tracer != null) {
-                if (headerExtractor != null) {
-                    HeaderExtractorBridge headerExtractorBridge = HeaderExtractorBridge.instance();
-                    headerExtractorBridge.setGetHeaderMethodHandle(getFirstHeader);
-                    transaction = tracer.startChildTransaction(headerExtractor, headerExtractorBridge, clazz.getClassLoader());
-                } else if (headersExtractor != null) {
-                    HeadersExtractorBridge headersExtractorBridge = HeadersExtractorBridge.instance();
-                    headersExtractorBridge.setGetHeaderMethodHandle(getAllHeaders);
-                    transaction = tracer.startChildTransaction(headersExtractor, headersExtractorBridge, clazz.getClassLoader());
+                if (headersExtractor != null) {
+                    HeadersExtractorBridge headersExtractorBridge = HeadersExtractorBridge.get(getFirstHeader, getAllHeaders);
+                    transaction = tracer.startChildTransaction(new HeadersExtractorBridge.Extractor(headerExtractor, headersExtractor), headersExtractorBridge, clazz.getClassLoader());
```

Comment:
Why not reusing the same instance? See how I changed in the merge code, let me know if you see a problem with that.